### DISCUSSION TITLE
Anonymous access with built-in database enabled not working

### DISCUSSION THREAD

**Discussion by cyfinfaza** [emqx/emqx#8651] [Q&A] [answered]
Hi,
I just installed EMQX 5.0.4 on Ubuntu Server 20.04. 

After configuring the built-in database authentication, I can no longer connect to the MQTT server without specifying a username and password (returns unauthorized). I have set `allow_anonymous=true` in `/etc/emqx/emqx.conf`, but it still doesn't work. 

My goal is to have anyone be able to connect anonymously and subscribe to topics, but only authenticated users can publish. How can I configure EMQX this way?

Thanks

**Comment by zmstone** [comment]
There is a enable_authn flag for each mqtt listener which can be set to false. Then all users will be able to connect.

for publish permission, you’ll need to configure authorization (ACL) rules instead.

**Reply by cyfinfaza** [reply]
This seems to work, except now it doesn't check passwords anymore. For example, I can login as admin with a random password and it still lets me publish messages, but it doesn't let me publish if I specify anything else for the username (regardless of the password).  
In the dashboard > Access Control > Authorization > Built-in Database I have it set to deny publish on # and allow subscribe on # for all users. (and admin has superuser and pub and sub on #)
![image](https://user-images.githubusercontent.com/42043351/182699895-6163f5e1-64ed-417f-bde0-b8a62753fa55.png)

**Reply by zmstone** [reply]
I see.
Will investigate how to allow anonymous clients to connect like 4.x

**Reply by cyfinfaza** [reply]
Thanks
If you would like any additional logs, config, etc. info from my instance let me know

**Comment by zmstone** [comment, marked as answer]
Hi again @cyfinfaza 

The `allow_anonymous` config has been removed from 5.0
Reason: this config caused confusion to the majority of the users, and poses a potential security loophole because `allow_anonymous` is by default `true`, meaning anonymous clients are allowed even with authn plug-ins enabled.

The new logic in 5.0 is:

* If there is not any authenticator configured (and enabled), all clients a allowed to connect
* If there is at leat one authenticator enabled, all clients has to be authenticated.
* Authenticator (chain) can be configured per-listener.
* A "shortcut" per-listener flag is added to explicitly disable auth for the listener (as if configured with empty authenticator chain)

In your case, there seem to be mixed users, you will need to create another listener with auth disabled, so the anonymous clients can connect to this dedicated listener.
For the clients which are intended to be authenticated e.g. for the purpose of `superuser` recognition etc. they should connect to a listener with auth enabled.

**Reply by cyfinfaza** [reply]
I see. Thanks for the info and the quick response.

**Reply by mauriceatkinson** [reply]
I am being dumb here but when trying to create a new listener to allow anonymous clients to connect to broker i cannot find any setting related to disabling of auth.  I am assuming this is in the add listener.  Am I looking in the wrong place?  Direction appreciated here.
